### PR TITLE
Bugfix: Clarification of Amounts in tables (fixes #1861)

### LINKED ISSUE #1861: UI Improvement needed:  Transaction history added/substracted amounts unclear
Receiving, sending and self transfers all show up as white amounts:
image

It is completely unclear what is added or subtracted from the wallet.  I propose a more typical (like electrum or bitcoin core) style formatting of the numbers:
image

1. Self-transfers should show the fees lost.  
2. added amounts should be in default color (optional is a +)
3. sent/subtracted amounts should be red to be consistent with any sensible book-keeping formatting

I think this is extremely important for readability. @k9ert , @moneymanolis  What do you think?

### PR DESCRIPTION
**Pull request by k9ert**
This PR addresses criticism on how amounts are shown and fixes some edge-cases around that and does some refactoring around txlists. In detail:
* Outflowing value is shown as a negative number
* Outflow includes the fee.
* If Txs have inputs which doesn't belong to the wallet, wrong numbers were shown. This is fixed now.
* There is also a clear-cache button in the settings of each wallet.
* Some refactorings
image

Before (history): 
image
After (history):
image

Before(utxo)
image
After (utxo):
image

### CODE REVIEW

File: src/cryptoadvance/specter/server_endpoints/wallets/wallets.py [hunk 1]
```diff
@@ -894,3 +894,22 @@ def settings(wallet_alias):
...
     )
+
+
+@wallets_endpoint.route("/wallet/<wallet_alias>/settings/clearcache", methods=["POST"])
+@login_required
+def settings_clearcache(wallet_alias):
+    wallet: Wallet = app.specter.wallet_manager.get_by_alias(wallet_alias)
+    error = None
+    wallet.clear_cache()
+    flash("Cache cleared successfully!")
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/server_endpoints/wallets/wallets.py:905]: Not so clear to the user which cache this is referring to.

**Comment by moneymanolis** [comment]
> There is also a clear-cache button in the settings of each wallet.

Is this pushed already, can't see anything in this regard.

File: src/cryptoadvance/specter/templates/includes/tx-table.html [hunk 2]
```diff
@@ -776,6 +776,11 @@ <h1>{{ _("Export transactions to CSV") }}</h1>
         */
         attributeChangedCallback(attrName, oldValue, newValue) {
             if (attrName != "selected-coins") {
+                if (this.getAttribute('type') == "txlist") {
+                    this.amountHeader.innerText="Flow Amount"
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/templates/includes/tx-table.html:780]:
```suggestion
                    this.amountHeader.innerText="Amount \u0394"
```

File: src/cryptoadvance/specter/templates/includes/tx-table.html [hunk 3]
```diff
@@ -776,6 +776,11 @@ <h1>{{ _("Export transactions to CSV") }}</h1>
...
+                } else {
+                    this.amountHeader.innerText = "Value Amount"
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/templates/includes/tx-table.html:782]:
```suggestion
                    this.amountHeader.innerText = "Amount"
```

File: docs/architecture.md [hunk 4]
```diff
@@ -28,4 +28,17 @@ This is also a pattern used for plugins.
...
 
-We now try to implement bigger chunks of functionality in plugins. Maybe, even core functionality might be implemented in "core plugins" in the future. Internal plugins are placed in `src/specterext`. But, plugins can also live in there own repos, have their own release lifecycle and be used by Specter like any other dependency. For more information about plugins see [Third Party Service Integrations](./extensions.md) where we also discuss the nuances between plugins and extensions.
\ No newline at end of file
+We now try to implement bigger chunks of functionality in plugins. Maybe, even core functionality might be implemented in "core plugins" in the future. Internal plugins are placed in `src/specterext`. But, plugins can also live in there own repos, have their own release lifecycle and be used by Specter like any other dependency. For more information about plugins see [Third Party Service Integrations](./extensions.md) where we also discuss the nuances between plugins and extensions.
+
+# The Crypto Engine
```
Comment (moneymanolis) [inline docs/architecture.md:33]: Not sure whether this is the right expression ...

File: src/cryptoadvance/specter/templates/wallet/settings/wallet_settings.jinja [hunk 5]
```diff
@@ -322,8 +322,26 @@
...
+			<h2 class="subtitle">{{ _("Clear Cache") }}</h2>
+			<div class="row center">
+				<form action="./clearcache" method="POST" class="padded" style="max-width: 500px">
+					<input type="hidden" class="csrf-token" name="csrf_token" value="{{ csrf_token() }}"/>
+					<button type="submit" name="action" value="clearcache" class="btn" style="max-width: 130px; margin-left: 20px">{{ _("Clear Cache") }}
+						<tool-tip width="200px">
+							<span slot="paragraph">
+								{{ _('Each wallet maintains a cache which stores the details of transactions') }}<br>
+								{{ _('in order to speed up loading of the tx-list. ') }}<br>
+								{{ _('Clearing that Cache is not dangerous and possible any time.') }}<br>
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/templates/wallet/settings/wallet_settings.jinja:335]:
```suggestion
								{{ _('Clearing that cache is not deleting any crucial data and possible any time.') }}<br>
```

File: src/cryptoadvance/specter/templates/wallet/settings/wallet_settings.jinja [hunk 6]
```diff
@@ -322,8 +322,26 @@
...
+								{{ _('Expect a bit longer loading time after you\'ve cleared the cache') }}<br><br>
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/templates/wallet/settings/wallet_settings.jinja:336]:
```suggestion
								{{ _('However, expect a longer loading time after clearing the cache') }}<br><br>
```

File: src/cryptoadvance/specter/server_endpoints/wallets/wallets.py [hunk 7]
```diff
@@ -882,7 +813,164 @@ def settings(wallet_alias):
...
+    )
+
+
+@wallets_endpoint.route("/wallet/<wallet_alias>/settings/clearcache", methods=["POST"])
+@login_required
+def settings_clearcache(wallet_alias):
+    wallet: Wallet = app.specter.wallet_manager.get_by_alias(wallet_alias)
+    error = None
+    wallet.clear_cache()
+    flash("Cache cleared successfully!")
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/server_endpoints/wallets/wallets.py:973]:
```suggestion
    flash("Cache with transactions cleared successfully!")
```

File: src/cryptoadvance/specter/services/extension_gen.py [hunk 8]
```diff
@@ -113,6 +113,7 @@ def generate_basics(self):
         self.render(f"{package_path}/templates/dummy/index.jinja")
         if self.devicename:
             self.render(f"{package_path}/devices/devicename.py")
+            self.render(f"{package_path}/devices/__init__.py")
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/services/extension_gen.py:116]: Is this intended to be here?

File: src/cryptoadvance/specter/templates/includes/tx-table.html (hunk 2 above)
Comment (relativisticelectron) [reply src/cryptoadvance/specter/templates/includes/tx-table.html:780]: Both electrum and bitcoin core call this "Amount".

File: src/cryptoadvance/specter/templates/includes/tx-table.html [hunk 9]
```diff
@@ -776,6 +776,11 @@ <h1>{{ _("Export transactions to CSV") }}</h1>
...
+                }
```
[the remark this answers is not in the document]
Comment (k9ert) [reply src/cryptoadvance/specter/templates/includes/tx-table.html:783]: accepted!

File: docs/architecture.md (hunk 4 above)
Comment (k9ert) [reply docs/architecture.md:33]: Not sure what you're referring to. What's wrong here? I added `src/cryptoadvance/specterext` but apart from that?!

File: src/cryptoadvance/specter/services/extension_gen.py (hunk 8 above)
Comment (k9ert) [reply src/cryptoadvance/specter/services/extension_gen.py:116]: Yes, tiny side-fix.

File: src/cryptoadvance/specter/server_endpoints/wallets/wallets.py (hunk 1 above)
Comment (k9ert) [reply src/cryptoadvance/specter/server_endpoints/wallets/wallets.py]: The tooltip looks like this:
image

**Comment by k9ert** [comment]
> With the PR checked out, my wallets are not loading completely (works fine on master):
Can't reproduce. Can you check that again?

**Review by moneymanolis — CHANGES_REQUESTED** [review]
Export of transaction history (completely empty CSV) and UTXOs (only headers) is not working.

**Comment by moneymanolis** [comment]
> > With the PR checked out, my wallets are not loading completely (works fine on master):
> 
> Can't reproduce. Can you check that again?

Works now.

**Comment by k9ert** [comment]
as predicted @moritzwietersheim :
```

[  ERROR] in      controller: Traceback (most recent call last):
  File "/tmp/cirrus-ci-build/.env/lib/python3.10/site-packages/flask/app.py", line 1523, in full_dispatch_request
    rv = self.dispatch_request()
  File "/tmp/cirrus-ci-build/.env/lib/python3.10/site-packages/flask/app.py", line 1509, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/tmp/cirrus-ci-build/.env/lib/python3.10/site-packages/flask_login/utils.py", line 269, in decorated_view
    return func(*args, **kwargs)
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/server_endpoints/wallets/wallets_api.py", line 397, in txlist
    txlist = wallet.txlist(
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/wallet.py", line 1083, in txlist
    self.fetch_transactions()
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/wallet.py", line 395, in fetch_transactions
    unconfirmed_selftransfers = [
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/wallet.py", line 398, in <listcomp>
    if self._transactions[txid].category == "selftransfer"
AttributeError: 'LTxItem' object has no attribute 'category'
[WARNING] in          txlist: this can probably be ignored: Failed to rewind the proof
[  ERROR] in      controller: Uncaught exception: 'LTxItem' object has no attribute 'category'
```
If it's cheap for me to fix that, i'll do it.

**Comment by k9ert** [comment]
Ok, Liquid fixes were not that difficult. At least not the use-cases qhich the tests cover. Should be all green now!

**Comment by moneymanolis** [comment]
> Export of transaction history (completely empty CSV) and UTXOs (only headers) is not working.

Export of transaction history works now but export of UTXOs is still only headers.

**Comment by moneymanolis** [comment]
The csv cache seems to be overwritten this is how it looks like, is this all as intended? @k9ert 

```
txid,fee,blockhash,blockheight,time,blocktime,bip125-replaceable,conflicts,vsize,address,category,flow_amount,utxo_amount,ismine
1a5206140a136c6e4d1901923d7fa765918904bb235bdae43de336e92507304a,,132aae8511edbd3f6ac52c6e253e1dade0eebbe1635766cf5a66b955399a7772,735,1620068439,,no,[],,"['bcrt1q6x05wm2r8ey6hztcz2l5quze7svkemjrms98he', 'OP_RETURN 6a24aa21a9edcf8fb925b7f34943bf39c7d5bcd1ba9378fd6bd936f3e0c7dfc68443623c2e3b']",generate,3.12500208,,True
1e488de679b96d87c5c6377fb81884d0b3695f417305de4d68a3a634129cdb1d,,174ecbe9cc62cbc2b51312a6d30b79234106add4a186b511c7e51c49bdc2b610,736,1622573532,,no,[],,"['bcrt1qp3e6thsy9e538llhj4fgze5m5c7usz8qz3tqdc', 'bcrt1q0ktwqaptyg6gveg03lg8q9hs6qxwtlkrwd5ysv']",send,-100.0003425,,True
2cba28d71bb1693e3c29f59d424fa23cf9ad113dfb905c39ba711cffd88a6f5a,,666f308fa50539f8f030bcbd338a1ea20af3f06d39a014fc7b1281ac7b2e5f09,746,1622573915,,no,[],,"['bcrt1qn0maknxzxpyw04xmaelquf2f079wmfpkpw8x4v', 'bcrt1qwkeamkcyttskdm46jdkye72fwm73ycc9lnlg8e']",receive,0.001,,True
fd8f82a067c4bf9115ddba25054ee0e76e08825d1630a8d3ea37e48091565ab0,,5f80aa6b1e346545ba6ab3c68f47d12375ec5ef789a0fe60f87ce3abfa3bf40b,747,1636826971,,no,[],,bcrt1q78lx4k2u92f2w0nyc88qxlcfga55spdgma8868,selftransfer,-1.4100000003125501e-06,,True
6fafeb80c4f8797c86519b9b42f9056f671b2c5f28e62660a1d86ee3cf52ba11,,5f80aa6b1e346545ba6ab3c68f47d12375ec5ef789a0fe60f87ce3abfa3bf40b,747,1636827151,,no,[],,"['bcrt1qtcrknq74ju8fhn48f08g246wcdea3ag8p2q0kc', 'OP_RETURN 6a24aa21a9ed49618264fd14f9d625bc87280494386150154140ace453869f556e285b4df18e']",generate,3.12500141,,True
6b80e12af1acb544c703ad6bcfefdd5eae729127b064d300c9029f0c7e3dfbd1,,7db38b0a4f3fa3048c1b9528acf327e65c8f4338ce045d80a04a0a5102d15d2e,748,1636827151,,no,[],,"['bcrt1qtcrknq74ju8fhn48f08g246wcdea3ag8p2q0kc', 'OP_RETURN 6a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf9']",generate,3.125,,True
2812cef72a307ab48adf94a238fd5d73210b7240db7c76c5c587a02aca8cfd2a,,517858df7200f7e5c9882fbebcbf2d04cd8c05cf2bc26e9d2b9485d21b9ddce0,749,1636827151,,no,[],,"['bcrt1qtcrknq74ju8fhn48f08g246wcdea3ag8p2q0kc', 'OP_RETURN 6a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf9']",generate,3.125,,True
```

**Comment by moneymanolis** [comment]
UTXO list export works now!

**Review by moneymanolis — APPROVED** [review]

**Comment by k9ert** [comment]
As discussed in the review, the case for 2 outputs in an incoming transaction into the same wallet myght cause headache with coin-selection. So i reimplemented the whole method the other way around:
[f1fb38e](https://github.com/cryptoadvance/specter-desktop/pull/2026/commits/f1fb38e82d639d3b3c8746cbf83327b629aa4a2f)

Now, if there are two utxos with the same `txid` but two `vout`, this will properly result in two `WalletAwareTxItem` both with the same `txid` but different `vout`.
CC @moneymanolis

**Pull request merged**